File: Medium/2025-08-28-3446-Sort-Matrix-by-Diagonals/solution.java
java
import java.util.*;

class Solution {
    public int[][] diagonalSort(int[][] grid) {
        int n = grid.length;
        Map<Integer, PriorityQueue<Integer>> diagMap = new HashMap<>();

        // Collect elements in diagonals for bottom-left triangle
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                diagMap.computeIfAbsent(i - j, k -> new PriorityQueue<>(Collections.reverseOrder())).offer(grid[i][j]);
            }
        }

        // Replace elements in bottom-left triangle with sorted values
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                grid[i][j] = diagMap.get(i - j).poll();
            }
        }

        diagMap.clear();

        // Collect elements in diagonals for top-right triangle
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                diagMap.computeIfAbsent(j - i, k -> new PriorityQueue<>()).offer(grid[i][j]);
            }
        }

        // Replace elements in top-right triangle with sorted values
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                grid[i][j] = diagMap.get(j - i).poll();
            }
        }

        return grid;
    }
}

File: Medium/2025-08-28-3446-Sort-Matrix-by-Diagonals/solution.py
python
class Solution:
    def diagonalSort(self, grid: List[List[int]]) -> List[List[int]]:
        n = len(grid)
        # Sort bottom-left diagonals
        for col in range(n):
            diag = []
            r, c = 0, col
            while r < n and c >= 0:
                diag.append(grid[r][c])
                r += 1
                c -= 1
            diag.sort(reverse=True)
            r, c = 0, col
            for val in diag:
                grid[r][c] = val
                r += 1
                c -= 1
        
        # Sort top-right diagonals
        for row in range(1, n):
            diag = []
            r, c = row, n - 1
            while r < n and c >= 0:
                diag.append(grid[r][c])
                r += 1
                c -= 1
            diag.sort()
            r, c = row, n - 1
            for val in diag:
                grid[r][c] = val
                r += 1
                c -= 1
        
        return grid